### ISSUE TITLE
Datasource Connection issues while migrating database from 17.0.1 to 21.0.0

### ISSUE THREAD

**Issue by gjain-sag** [keycloak/keycloak#19453]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

account/api

### Describe the bug

I have a Keycloak (wildfly) instance with version 17.0.1 having ~600 realms and while migrating the same to keycloak-21.0.1, keycloak instance fails to start with below exception :

2023-03-30 23:22:37,405 DEBUG [org.hibernate.SQL] (main) select cliententi0_.ID as col_0_0_ from CLIENT cliententi0_ where cliententi0_.CLIENT_ID=? and cliententi0_.REALM_ID=?
2023-03-30 23:22:37,406 DEBUG [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (main) could not prepare statement [select cliententi0_.ID as col_0_0_ from CLIENT cliententi0_ where cliententi0_.CLIENT_ID=? and cliententi0_.REALM_ID=?]: java.sql.SQLException: Connection is closed
        at io.agroal.pool.wrapper.ConnectionWrapper.lambda$static$0(ConnectionWrapper.java:52)
        at com.sun.proxy.$Proxy62.prepareStatement(Unknown Source)
        at io.agroal.pool.wrapper.ConnectionWrapper.prepareStatement(ConnectionWrapper.java:659)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:149)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:176)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:151)
        at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:2122)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2059)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037)
        at org.hibernate.loader.Loader.doQuery(Loader.java:956)
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:357)
        at org.hibernate.loader.Loader.doList(Loader.java:2868)
        at org.hibernate.loader.Loader.doList(Loader.java:2850)
        at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2682)
        at org.hibernate.loader.Loader.list(Loader.java:2677)
        at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:540)
        at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:400)
        at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:219)
        at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1459)
        at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1649)
        at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617)
        at org.hibernate.query.Query.getResultList(Query.java:165)
        at org.keycloak.models.jpa.JpaRealmProvider.getClientByClientId(JpaRealmProvider.java:760)
        at org.keycloak.storage.ClientStorageManager.getClientByClientId(ClientStorageManager.java:149)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.getClientByClientId(RealmCacheSession.java:1209)
        at org.keycloak.models.cache.infinispan.RealmAdapter.getClientByClientId(RealmAdapter.java:815)
        at org.keycloak.models.utils.KeycloakModelUtils.lambda$getClientScopeByName$13(KeycloakModelUtils.java:819)
        at java.base/java.util.Optional.orElseGet(Optional.java:369)
        at org.keycloak.models.utils.KeycloakModelUtils.getClientScopeByName(KeycloakModelUtils.java:819)
        at org.keycloak.protocol.oidc.OIDCLoginProtocolFactory.addAcrClientScope(OIDCLoginProtocolFactory.java:372)
        at org.keycloak.services.migration.DefaultMigrationProvider.addOIDCAcrClientScope(DefaultMigrationProvider.java:107)
        at org.keycloak.migration.migrators.MigrateTo18_0_0.migrateRealm(MigrateTo18_0_0.java:60)
        at org.keycloak.migration.migrators.MigrateTo18_0_0.lambda$migrate$0(MigrateTo18_0_0.java:46)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at org.hibernate.query.spi.StreamDecorator.forEach(StreamDecorator.java:153)
        at org.keycloak.utils.ClosingStream.forEach(ClosingStream.java:128)
        at org.keycloak.migration.migrators.MigrateTo18_0_0.migrate(MigrateTo18_0_0.java:46)
        at org.keycloak.storage.datastore.LegacyMigrationManager.migrate(LegacyMigrationManager.java:135)
        at org.keycloak.migration.MigrationModelManager.migrate(MigrationModelManager.java:33)
        at 
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:93)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)

2023-03-30 23:22:37,407 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (main) SQL Error: 0, SQLState: null
2023-03-30 23:22:37,407 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (main) Connection is closed
2023-03-30 23:22:37,409 DEBUG [org.hibernate.resource.jdbc.internal.ResourceRegistryStandardImpl] (main) Exception clearing maxRows/queryTimeout [Statement is closed]
2023-03-30 23:22:37,409 DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (main) Initiating JDBC connection release from afterTransaction
2023-03-30 23:22:37,409 DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (main) Initiating JDBC connection release from afterTransaction
2023-03-30 23:22:37,424 DEBUG [org.hibernate.engine.transaction.internal.TransactionImpl] (main) On TransactionImpl creation, JpaCompliance#isJpaTransactionComplianceEnabled == false
2023-03-30 23:22:37,425 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (main) JtaTransactionWrapper rollback
2023-03-30 23:22:37,426 WARN  [com.arjuna.ats.arjuna] (main) ARJUNA012077: Abort called on already aborted atomic action 0:ffff0a3c1e68:a4bd:6425cb36:0
2023-03-30 23:22:37,426 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (main) JtaTransactionWrapper end
2023-03-30 23:22:37,460 DEBUG [org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl] (main) HHH000420: Closing un-released batch
2023-03-30 23:22:37,462 DEBUG [org.infinispan.manager.DefaultCacheManager] (main) Stopping cache manager XXXXXX-39174
2023-03-30 23:22:37,463 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache authenticationSessions on XXXXX-39174
2023-03-30 23:22:37,463 DEBUG [org.infinispan.topology.LocalTopologyManagerImpl] (main) Node XXXXX-39174 leaving cache authenticationSessions
2023-03-30 23:22:37,464 DEBUG [org.infinispan.partitionhandling.impl.PreferAvailabilityStrategy] (main) The last node of cache authenticationSessions left
2023-03-30 23:22:37,468 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache realms on keycloak
2023-03-30 23:22:37,474 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache sessions on XXXXX-39174
2023-03-30 23:22:37,474 DEBUG [org.infinispan.topology.LocalTopologyManagerImpl] (main) Node XXXXX-39174 leaving cache sessions
2023-03-30 23:22:37,474 DEBUG [org.infinispan.partitionhandling.impl.PreferAvailabilityStrategy] (main) The last node of cache sessions left
2023-03-30 23:22:37,475 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache authorizationRevisions on keycloak
2023-03-30 23:22:37,476 DEBUG [org.infinispan.transaction.impl.TransactionTable] (main) Wait for on-going transactions to finish for 30 seconds.
2023-03-30 23:22:37,476 DEBUG [org.infinispan.transaction.impl.TransactionTable] (main) All transactions terminated
2023-03-30 23:22:37,476 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache clientSessions on XXXXX-39174
2023-03-30 23:22:37,476 DEBUG [org.infinispan.topology.LocalTopologyManagerImpl] (main) Node xXXXX-39174 leaving cache clientSessions
2023-03-30 23:22:37,476 DEBUG [org.infinispan.partitionhandling.impl.PreferAvailabilityStrategy] (main) The last node of cache clientSessions left
2023-03-30 23:22:37,478 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache keys on keycloak
2023-03-30 23:22:37,478 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache work on XXXXX-39174
202
org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.migrateModel(LegacyJpaConnectionProviderFactory.java:216)
2023-03-30 23:22:37,565 DEBUG [io.quarkus.runtime.Application] (main) Stopping application
2023-03-30 23:22:37,566 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2023-03-30 23:22:37,566 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: org.hibernate.HibernateException: Transaction was rolled back in a different thread!
2023-03-30 23:22:37,566 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Transaction was rolled back in a different thread!
2 

### Version

21.0.0

### Expected behavior

Keycloak 17.0.1 database should be migrated to keycloak 21.0.0 and keycloak instance should started 

### Actual behavior

Migration of Keycloak 17.0.1 (Wildfly) to Keycloak-21.0.0 (quarkus) on database with ~600 realms is failing to migrate and keycloak server fails to start.

### How to Reproduce?

1. Should have an Keycloak instance (17.0.1) with ~500 realms 
2. Install Keycloak 21.0.0 and point to Keycloak-17.0.1 database
3. Start server

**Comment by kkcmadhu** [comment]
very strage. i too see similar thing when upgrading to keyclok 21.0.2

```
        at org.hibernate.query.Query.getResultStream(Query.java:1140)
        at org.keycloak.models.jpa.JpaRealmProvider.getClientScopes(JpaRealmProvider.java:987)
        at org.keycloak.storage.ClientStorageManager.getClientScopes(ClientStorageManager.java:212)
.......
.....
.....

Caused by: java.sql.SQLException: Connection is closed
        at io.agroal.pool.wrapper.ConnectionWrapper.lambda$static$0(ConnectionWrapper.java:52)

```

**Comment by kkcmadhu** [comment]
Well.. it was quarkus transaction manager creating ruckus , this helped in tame the quarkus transaction manager
        - name: quarkus.transaction-manager.default-transaction-timeout
          value: "35M"

**Comment by kkcmadhu** [comment]
in my opinion when some one sets --spi-dblock-jpa-lock-wait-timeout, the  quarkus.transaction-manager.default-transaction-timeout  should also be overriden automatically unless a specific  configuration is added for  quarkus.transaction-manager.default-transaction-timeout with a differnt value.

**Comment by hmlnarik** [comment]
We need to ensure that if `quarkus.transaction-manager.default-transaction-timeout` is set, it is propagated to the value of lock timeouts: 
- `spi-dblock-jpa-lock-wait-timeout`
- `spi-mapStorage-hotRod-lockTimeout`
- `spi-mapStorage-jpa-lockTimeout`

**Comment by kkcmadhu** [comment]
> We need to ensure that if `quarkus.transaction-manager.default-transaction-timeout` is set, it is propagated to the value of lock timeouts:
> 
> * `spi-dblock-jpa-lock-wait-timeout`
> * `spi-mapStorage-hotRod-lockTimeout`
> * `spi-mapStorage-jpa-lockTimeout`

Oh thanks for the point, i will set them manually for now.
But, can you let me know how i can do it with environment variables, sould i use the convention
KC_SPI_DB_LOCK_JPA_LOCK_WAIT_TIMEOUT and so on?

**Comment by kkcmadhu** [comment]
fun part is if i just add  spi-dblock-jpa-lock-wait-timeout, it does not work..  seems to get around this issue quarkus.transaction-manager.default-transaction-timeout is required!

**Comment by boris-nu** [comment]
@kkcmadhu where did you add this?

Well.. it was quarkus transaction manager creating ruckus , this helped in tame the quarkus transaction manager
        - name: quarkus.transaction-manager.default-transaction-timeout
          value: "35M"
          
          
I am facing a similiar issues upgrading from 18.0 to 21.0.2 with +400 realms. KC is communicating with a PostgresDB of version 14.

**Comment by boris-nu** [comment]
@kkcmadhu thank you very much for your quick reply!!!  Just added it to the quarkus props..lets see if it runs through...any chance that you can share all of your env vars, or was this the trick to let the upgrade finish without any errors?

**Comment by boris-nu** [comment]
@kkcmadhu worked like a charm!!! You are a hero, saved my day and less hours of maintenance on production!!! This is currently not supported in keycloak to add it as env via compose?? Well did not work for me, only adding it to the quarkus.porperties...

**Comment by nitin-vavdiya** [comment]
Hi @boris-nu 
Can you please share enviorment variable that you have configured?
I am running keycloak in side the kubernate cluster.

**Comment by shawkins** [comment]
At least for JPA related issues https://github.com/keycloak/keycloak/issues/38649 and related changes have improved migration performance when there are a lot of realms. The should be considered initially resolved in KC 26.3.

With a sufficient number of realms it is still possible to exceed the default-transaction-timeout - that class of problems is to be addressed by https://github.com/keycloak/keycloak/issues/26889

On the db lock timeout the keycloak operator for example tries to avoid this situation by serially starting the instances. Only once an instance is ready will another start - this mostly sidesteps the need for the lock and helps prevent some issues with JGroups as well. 

@ahus1 @ryanemerson @pruivo and others - should this issue be added for consideration under https://github.com/keycloak/keycloak/issues/26889 or should be generally recommending serial startup behavior?

**Comment by pruivo** [comment]
@shawkins serial startup is better for JGroups/Infinispan.

On startup, JGroups reads and registers itself in the database; concurrent startup can originate the scenario where multiple instances read the database and it is empty and they won't connect to each other. Eventually, JGroups will detect the partition and fix itself.

**Comment by ahus1** [comment]
I've updated this issues's title to "The default database transaction timeout should not be applied to Liquibase or data migrations"

The current workaround seems to be setting `quarkus.transaction-manager.default-transaction-timeout` which is a Quarkus option that is not documented. 

Ideally we would have a longer default for all Liquibase schema migrations and data migrations via `MigrationTo*` classes. As people might want to configure this, it should be a be a first-class CLI option. 

Depending on how transaction timeouts would work: 
* There is a transaction for each Liquibase changeset, so any changeset must not exceed that duration
* There is a transaction spanning the migration of all realms, which is a good thing as we don't handle a situation where some realms are migrated while others are not. Maybe there is a way that the process can let the transaction manager know that it is making good progress, for example by pinging it after each realm. 

@pruivo - please investigate what can be done here. No need to dive into the implementation yet, let's first collect what's possible. 

**TL;DR**: It would be good to have two timeouts: One for migrations, another one for regular operations.

This is now part of making "Keycloak database instructions good enough to be production ready".

**Comment by shawkins** [comment]
From what I recall the transaction timeout from our top-level transaction will be enforced. In some places we suspend that one and start another for a subset of work, but the initial timeout will still be enforced even when suspended.

At the very least there is a containing transaction for the entire migration in QuarkusJpaConnectionProviderFactory.postInit. There is also already a runJobInTransactionWithTimeout method to control the transaction timeout on a per transaction basis.

**Comment by ahus1** [comment]
> From what I recall the transaction timeout from our top-level transaction will be enforced

There might be a loophole (still to be tested): The transaction reaper might only see transactions that connected to the JPA transaction. So if the parent transaction did not use JPA, maybe it will not be killed by the transaction reaper. At the same time the child transaction might continue if it was started with a longer transaction length. 

@pruivo - please investigate. Maybe a `Thread.sleep()` in one of the Migrations can give us some insights.

**Comment by shawkins** [comment]
> The transaction reaper might only see transactions that connected to the JPA transaction.

I don't believe so. The transaction timeout of all transactions, regardless of whether they are suspended is enforced.

Maybe @mmusgrov can offer some options wrt timeouts.

**Comment by ahus1** [comment]
Next step: 
* Try out with the Java Migrations to see if we can get it to work somehow, and when/how the transaction reaper kicks in. 
* We could have a "sleep" in the Migrations to try it out

**Comment by ahus1** [comment]
Related to #26889

**Comment by pruivo** [comment]
> > From what I recall the transaction timeout from our top-level transaction will be enforced
> 
> There might be a loophole (still to be tested): The transaction reaper might only see transactions that connected to the JPA transaction. So if the parent transaction did not use JPA, maybe it will not be killed by the transaction reaper. At the same time the child transaction might continue if it was started with a longer transaction length.
> 
> [@pruivo](https://github.com/pruivo) - please investigate. Maybe a `Thread.sleep()` in one of the Migrations can give us some insights.

@ahus1 It doesn't matter. I put a sleep inside `org.keycloak.models.utils.KeycloakModelUtils#suspendJtaTransaction()` and the transaction is still aborted by the `TransactionReaper`.
